[21.05] Pl 127635 refactor fc ceph

ctheune opened the pull request:
@flyingcircusio/release-managers

## Release process

Impact:

n/a

Changelog:

(internal ceph utility refactoring, no changelog entry needed?)

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)

Ensure proper argument handling and data structure handling when calling external programs. We refactored things to unify the different ways of calling external programs and at the same time fold various methods of interacting with CLI and JSON CLI programs into a homogenous internal API.

- [x] Security requirements tested? (EVIDENCE)

Pure internal refactoring, unit tests and hydra tests still work.

ctheune committed:
tests/ceph: improve output in failure situations

We noticed that failures during fc-ceph mon create went unnoticed
(due to using `execute` instead of `succeed`) and also that no real
output was provided. It appears that the &> redirection pattern isn't
properly supported here so we replaced it with the more typical 2>&1
pattern.

Also, avoid showing the journal output after creating the first mon
too early as it initially shows the (expected) failures of the
non-bootstrapped mon and that can be quite confusing.

ctheune committed:
fc-ceph: improve error reporting for multi-osd management tasks

osnyx commented on pkgs/fc/ceph/fc/ceph/util.py:
```diff
@@ -0,0 +1,102 @@
+        result = getattr(self.runner, name)(*args, **kw)
+        result = json.loads(result)
+        return result
+
```
For an easier superficial code structure, I suggest marking the tool-specific overrides with a comment to make the role of these member functions obvious at first glance. 

```suggestion

# tool-specific overrides of JSON formatting and extraction

```

ctheune replied:
thanks. added more docs as suggested.

osnyx commented on pkgs/fc/ceph/fc/ceph/manage.py:
```diff
@@ -420,18 +344,13 @@ def deactivate(self, flush=True):
+            run.ceph_osd(
+                # fmt: off
+                "-i", str(self.id),
+                "--flush-journal", "--osd-data",
```
The grouping of arguments is a bit odd here. I'd expect argument name and argument value to be on the same line (e.g `"--osd-data", self.datadir,`).

ctheune replied:
good catch!

osnyx commented on pkgs/fc/ceph/fc/ceph/util.py:
```diff
@@ -0,0 +1,115 @@
+        result = json.loads(result)
+        return result
+
+    # Tool-specific overrides to ensure that the invoed tools
```
```suggestion
    # Tool-specific overrides to ensure that the involved tools
```

typo-alarm 🚨

ctheune committed:
fc-ceph: refactor how we call external commands

We noticed that black continuously formats external command calls
in a basically unreadable fashion and started introducing fmt:off
blocks.

At the same time we had a multitude of different run helpers
and even residual calls not leveraging JSON interfaces from back when
this was running on Gentoo.

I created a runner structure that simplifies the Python code on the
application side and also provides a structured way to interact with
JSON CLIs.
